### PR TITLE
Extended query feature

### PR DESCRIPTION
**Pull request by EmadAlblueshi**
The PR includes below :

- Extended Query (Prepared Statement)
- New data type added
  1. JSON  for both of `io.vertx.core.json.JsonObject` and `io.vertx.core.json.JsonArray`
  2. JSONB for both of `io.vertx.core.json.JsonObject` and `io.vertx.core.json.JsonArray`

### CODE REVIEW

File: src/main/java/io/vertx/pgclient/codec/Column.java [hunk 1]
```diff
@@ -9,17 +9,17 @@
 public class Column {
 
   private final String name;
-  private final int type;
-  private final short format;
+  private final int dataType;
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/Column.java]: I think we should use the existing `DataType` enum

File: src/main/java/io/vertx/pgclient/codec/Column.java [hunk 2]
```diff
@@ -9,17 +9,17 @@
...
+  private final short dataFormat;
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/Column.java]: I think we should use the existing `DataFormat` enum

File: src/main/java/io/vertx/pgclient/codec/DataFormat.java [hunk 3]
```diff
@@ -1,10 +1,10 @@
-package io.vertx.pgclient.codec.decoder.message;
+package io.vertx.pgclient.codec;
 
 /**
  * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
  */
 
-public class ColumnFormat {
+public class DataFormat {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/DataFormat.java]: should make an enum with a `short` code

File: src/main/java/io/vertx/pgclient/codec/DataType.java [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
  *
  * PostgreSQL <a href="https://github.com/postgres/postgres/blob/master/src/include/catalog/pg_type.h">object
- * identifiers (OIDs)</a> for column data types
+ * identifiers (OIDs)</a> for data types
  *
  * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
  */
 
-public class ColumnType {
+public class DataType {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/DataType.java]: should make an enum with an `int` code

File: src/main/java/io/vertx/pgclient/codec/decoder/message/BindComplete.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package io.vertx.pgclient.codec.decoder.message;
+
+import io.vertx.pgclient.codec.Message;
+
+/**
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class BindComplete implements Message {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/decoder/message/BindComplete.java]: should make a singleton with private constructor and a `public static final BindComplete INSTANCE`

File: src/main/java/io/vertx/pgclient/codec/decoder/message/CloseComplete.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package io.vertx.pgclient.codec.decoder.message;
+
+import io.vertx.pgclient.codec.Message;
+
+/**
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class CloseComplete implements Message {
+
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/decoder/message/CloseComplete.java]: should make a singleton with private constructor and a `public static final CloseComplete INSTANCE`

File: src/main/java/io/vertx/pgclient/codec/decoder/message/EmptyQueryResponse.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ *
+ * <p>
+ * Identifies the message as a response to an empty query string. (This substitutes for {@link CommandComplete})
+ *
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class EmptyQueryResponse implements Message {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/decoder/message/EmptyQueryResponse.java]: should make a singleton with private constructor and a `public static final EmptyQueryResponse INSTANCE`

File: src/main/java/io/vertx/pgclient/codec/decoder/message/NoData.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package io.vertx.pgclient.codec.decoder.message;
+
+
+import io.vertx.pgclient.codec.Message;
+
+/**
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class NoData implements Message {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/decoder/message/NoData.java]: should make a singleton with private constructor and a `public static final NoData INSTANCE`

File: src/main/java/io/vertx/pgclient/codec/encoder/message/PortalSuspended.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * <p>
+ * The appearance of this message tells the frontend that another {@link Execute} should be issued against the
+ * same portal to complete the operation. The {@link CommandComplete} message indicating completion of the source
+ * SQL command is not sent until the portal's execution is completed
+ *
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class PortalSuspended implements Message {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/encoder/message/PortalSuspended.java]: should make a singleton with private constructor and a `public static final PortalSuspended INSTANCE`

File: src/main/java/io/vertx/pgclient/codec/encoder/message/Sync.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+ * and only one {@link ReadyForQuery} sent for each of this message.
+ *
+ * <p>
+ * Note this message does not cause a transaction block opened with BEGIN to be closed. It is possible to detect this
+ * situation in {@link ReadyForQuery#getTransactionStatus()} that includes {@link TransactionStatus} information.
+ *
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class Sync implements Message {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/encoder/message/Sync.java]: should make a singleton with private constructor and a `public static final Sync INSTANCE`

File: src/main/java/io/vertx/pgclient/codec/util/Util.java [hunk 11]
```diff
@@ -45,4 +58,43 @@ public static void writeCString(ByteBuf dst, byte[] bytes) {
     dst.writeBytes(bytes, 0, bytes.length);
     dst.writeByte(0);
   }
+
+  public static byte[][] paramValues(List<Object> parameters) {
+    byte[][] params = new byte[parameters.size()][];
+    for (short c = 0; c < parameters.size(); ++c) {
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/codec/util/Util.java]: just use `int` the VM can better optimize it

File: src/main/java/io/vertx/pgclient/PostgresConnection.java [hunk 12]
```diff
@@ -3,15 +3,20 @@
...
  * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
  */
+
 public interface PostgresConnection {
 
   void exceptionHandler(Handler<Throwable> handler);
 
   void execute(String sql, Handler<AsyncResult<Result>> handler);
 
+  void execute(String sql, List<Object> params, Handler<AsyncResult<Result>> handler);
```
Comment (vietj) [inline src/main/java/io/vertx/pgclient/PostgresConnection.java]: I think we should rename it to prepareAndExecute as this is not the same than the simple `execute` and provide overloading to ease API:

- `prepareAndExecute(String sql, Object param1, Handler<AsyncResult<Result>>)`
- `prepareAndExecute(String sql, Object param1, Object param2, Handler<AsyncResult<Result>>)`

up to a few ones (5 or 6)

File: src/test/java/io/vertx/pgclient/PgTestBase.java [hunk 13]
```diff
@@ -548,14 +551,233 @@ public void testTimestampTzTextDataType(TestContext ctx) {
...
+          async.complete();
+        }));
+    }));
+  }
+
+  @Test
+  public void testExtendedSelect(TestContext ctx) {
+    Async async = ctx.async();
+    PostgresClient client = PostgresClient.create(vertx, options);
+    connector.accept(client, ctx.asyncAssertSuccess(conn -> {
```
Comment (vietj) [inline src/test/java/io/vertx/pgclient/PgTestBase.java]: use new overloaded method for better readability

File: src/test/java/io/vertx/pgclient/PgTestBase.java [hunk 14]
```diff
@@ -548,14 +551,233 @@ public void testTimestampTzTextDataType(TestContext ctx) {
...
+        }));
+    }));
+  }
+
+  @Test
+  public void testExtendedUpdate(TestContext ctx) {
+    Async async = ctx.async();
+    PostgresClient client = PostgresClient.create(vertx, options);
+    connector.accept(client, ctx.asyncAssertSuccess(conn -> {
+      conn.prepareAndExecute("UPDATE Fortune SET message = $1 WHERE id = $2", Arrays.asList("Whatever", 20),
```
Comment (vietj) [inline src/test/java/io/vertx/pgclient/PgTestBase.java]: use new overloaded method for better readability

**Comment by EmadAlblueshi** [comment]
@vietj Done

**Comment by vietj** [comment]
thanks

**Comment by EmadAlblueshi** [comment]
@vietj More than welcome :)

**Pull request merged**